snuba/datasets/storages/storage_key.py

```diff
@@ -0,0 +1,81 @@
+from __future__ import annotations
+
+from functools import lru_cache
+from typing import Any
+
+HARDCODED_STORAGE_KEYS = {
+    "DISCOVER": "discover",
+    "EVENTS": "events",
+    "EVENTS_RO": "events_ro",
+    "ERRORS": "errors",
+    "ERRORS_RO": "errors_ro",
+    "GROUPEDMESSAGES": "groupedmessages",
+    "GROUPASSIGNEES": "groupassignees",
+    "METRICS_COUNTERS": "metrics_counters",
+    "ORG_METRICS_COUNTERS": "org_metrics_counters",
+    "METRICS_DISTRIBUTIONS": "metrics_distributions",
+    "METRICS_SETS": "metrics_sets",
+    "METRICS_RAW": "metrics_raw",
+    "OUTCOMES_RAW": "outcomes_raw",
+    "OUTCOMES_HOURLY": "outcomes_hourly",
+    "QUERYLOG": "querylog",
+    "SESSIONS_RAW": "sessions_raw",
+    "SESSIONS_HOURLY": "sessions_hourly",
+    "ORG_SESSIONS": "org_sessions",
+    "SPANS": "spans",
+    "TRANSACTIONS": "transactions",
+    "TRANSACTIONS_RO": "transactions_ro",
+    "TRANSACTIONS_V2": "transactions_v2",
+    "ERRORS_V2": "errors_v2",
+    "PROFILES": "profiles",
+    "FUNCTIONS": "functions",
+    "FUNCTIONS_RAW": "functions_raw",
+    "REPLAYS": "replays",
+    "ERRORS_V2_RO": "errors_v2_ro",
+}
+
+REGISTERED_STORAGE_KEYS: dict[str, str] = {}
+
+
+class _StorageKey(type):
+    def __getattr__(cls, attr: str) -> "StorageKey":
+        if attr not in HARDCODED_STORAGE_KEYS and attr not in REGISTERED_STORAGE_KEYS:
```

Comment:
The next step of this needs to be locking down the dependency injection of things. It feels like we have all these disparate modules with a mix of things built lazily and on import that all happen to be initialized in the right order. Feels like it's pretty brittle right now.